bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -1091,24 +1159,33 @@ private Header getHeaderForLogId(long entryLogId) throws IOException {
     }
 
     private BufferedReadChannel getChannelForLogId(long entryLogId) throws IOException {
-        BufferedReadChannel fc = getFromChannels(entryLogId);
-        if (fc != null) {
-            return fc;
+        BufferedReadChannel brc = getFromChannels(entryLogId);
+        if (brc != null) {
+            return brc;
         }
+
         File file = findFile(entryLogId);
         // get channel is used to open an existing entry log file
         // it would be better to open using read mode
         FileChannel newFc = new RandomAccessFile(file, "r").getChannel();
-        FileChannel oldFc = logid2FileChannel.putIfAbsent(entryLogId, newFc);
+        ReferenceCountedFileChannel oldFc =
+                logid2FileChannel.putIfAbsent(entryLogId, new ReferenceCountedFileChannel(newFc));
         if (null != oldFc) {
-            newFc.close();
-            newFc = oldFc;
+            // increment the refCnt
+            // double check to ensure the fileChannel is not closed due to refCnt down to 0.
+            if (oldFc.refCnt() > 0 && !oldFc.dead.get()) {
```

Comment:
A cache on another thread could have released between L1176 and L1177, invalidating the oldFc object.